crates/compose/src/lib.rs

```diff
@@ -28,18 +28,68 @@ use wac_graph::{CompositionGraph, NodeId};
 /// composition graph into a byte array and return it.
 pub async fn compose<'a, L: ComponentSourceLoader>(
     loader: &'a L,
-    component: &LockedComponent,
+    component: &L::Component,
 ) -> Result<Vec<u8>, ComposeError> {
     Composer::new(loader).compose(component).await
 }
 
+#[async_trait::async_trait]
+pub trait DependencyLike {
+    fn inherit(&self) -> InheritConfiguration;
+    fn export(&self) -> &Option<String>;
+}
+
+pub enum InheritConfiguration {
+    All,
+    Some(Vec<String>),
+}
+
+#[async_trait::async_trait]
+pub trait ComponentLike {
+    type Dependency: DependencyLike;
+
+    fn dependencies(
+        &self,
+    ) -> impl std::iter::ExactSizeIterator<Item = (&DependencyName, &Self::Dependency)>;
+    fn id(&self) -> &str;
+}
+
+#[async_trait::async_trait]
+impl ComponentLike for spin_app::locked::LockedComponent {
+    type Dependency = spin_app::locked::LockedComponentDependency;
+
+    fn dependencies(
+        &self,
+    ) -> impl std::iter::ExactSizeIterator<Item = (&DependencyName, &Self::Dependency)> {
+        self.dependencies.iter()
+    }
+
+    fn id(&self) -> &str {
+        &self.id
+    }
+}
+
+#[async_trait::async_trait]
+impl DependencyLike for spin_app::locked::LockedComponentDependency {
+    fn inherit(&self) -> InheritConfiguration {
+        match &self.inherit {
+            LockedInheritConfiguration::All => InheritConfiguration::All,
+            LockedInheritConfiguration::Some(cfgs) => InheritConfiguration::Some(cfgs.clone()),
+        }
+    }
+
+    fn export(&self) -> &Option<String> {
+        &self.export
+    }
+}
+
 /// This trait is used to load component source code from a locked component source across various embdeddings.
 #[async_trait::async_trait]
 pub trait ComponentSourceLoader {
-    async fn load_component_source(
-        &self,
-        source: &locked::LockedComponentSource,
-    ) -> anyhow::Result<Vec<u8>>;
+    type Component: ComponentLike<Dependency = Self::Dependency>;
+    type Dependency: DependencyLike;
+    async fn load_component_source(&self, source: &Self::Component) -> anyhow::Result<Vec<u8>>;
+    async fn load_dependency_source(&self, source: &Self::Dependency) -> anyhow::Result<Vec<u8>>;
```

Comment:
Are we implementing `ComponentSourceLoader` more than once? I'm a bit lost why we need the flexibility of defining the type of dependency and component instead of hard coding. Can you explain what this buys us?

Reply:
The current build system doesn't create a lockfile.  The current composition implementation depends on the locked types (because it runs at trigger time).  This allows us to implement composition on the raw AppManifest as well as on the LockedApp.